ARM builds fail

On 03/12/18 09:27, Jacob Gorm Hansen wrote:
> On Sunday, March 11, 2018 at 11:22:40 PM UTC+1, Field Van Zee wrote:
>> Jacob,
>>
>> I have corrected the typos in 1442d06. Now, this may not be the *only* error standing in your way, but it definitely needed to get fixed. Please give it a try.
>
> It now gets a bit further, but ultimately dies with:
>
> Compiling obj/cortexa57/frame/base/bli_func.o ('cortexa57' CFLAGS for framework code)
> frame/base/bli_cpuid.c: In function "bli_cpuid_query_id":
> frame/base/bli_cpuid.c:54:17: error: "VENDOR_INTEL" undeclared (first use in this function); did you mean "VENDOR_ARM"?
>   if ( vendor == VENDOR_INTEL )
>                  ^~~~~~~~~~~~
>                  VENDOR_ARM
> compilation terminated due to -Wfatal-errors.
> make[1]: *** [Makefile:462: obj/cortexa57/frame/base/bli_cpuid.o] Error 1

Jacob: I've committed 1a303174, which hopefully addresses some of the issues. Please try it out and let me know the outcome.

Thank you!
it seems to be working for cortexa57 now. However, for cortexa15 (or a9) I am still getting misnamed kernel errors, and I had to apply this Makefile fix:

diff --git a/config/cortexa9/make_defs.mk b/config/cortexa9/make_defs.mk
index 684879da..8989dbd2 100644
--- a/config/cortexa9/make_defs.mk
+++ b/config/cortexa9/make_defs.mk
@@ -67,10 +67,10 @@ endif
 
 CKOPTFLAGS     := $(COPTFLAGS)
 
-ifneq ($(CC_VENDOR),gcc)
+ifeq ($(CC_VENDOR),gcc)
 CVECFLAGS      := -march=armv7-a
 else
-$(error gcc is required for this configuration.)
+$(error gcc is required for this configuration. $(CC_VENDOR))
 endif
 
 # Store all of the variables here to new variables containing the
  
Here is the make error I am getting for cortexa15:

configure: configured to build within top-level directory of source distribution.
make[1]: warning: -jN forced in submake: disabling jobserver mode.
make[1]: Entering directory '/home/jacob/dev/simplecnn/blis'
make[1]: *** No rule to make target 'obj/cortexa15/kernels/armv7a/3/bli_cgemm_armv7a_asm_2x2.o', needed by 'lib/cortexa15/libblis.a'.  Stop.
make[1]: *** Waiting for unfinished jobs....

--- a/Makefile
+++ b/Makefile
@@ -432,7 +432,7 @@ endef
 # third argument: the kernel file suffix being considered.
 #$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.c $(BLIS_H_FLAT) $(MAKE_DEFS_MK_PATHS)
 define make-kernels-rule
-$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.$(3) $(MK_HEADER_FILES) $(MAKE_DEFS_MK_PATHS)
+$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.$(3) $(BLIS_H_FLAT) $(MAKE_DEFS_MK_PATHS)
 ifeq ($(BLIS_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
 	$(CC) $(call get-kernel-cflags-for,$(2)) -c $$< -o $$@
 else

--- a/Makefile
+++ b/Makefile
@@ -430,9 +430,9 @@ endef
 # first argument: a kernel set (name) being targeted (e.g. haswell).
 # second argument: the configuration whose CFLAGS we should use in compilation.
 # third argument: the kernel file suffix being considered.
-#$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.$(3) $(MK_HEADER_FILES) $(MAKE_DEFS_MK_PATHS)
+#$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.c $(BLIS_H_FLAT) $(MAKE_DEFS_MK_PATHS)
 define make-kernels-rule
-$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.c $(BLIS_H_FLAT) $(MAKE_DEFS_MK_PATHS)
+$(BASE_OBJ_KERNELS_PATH)/$(1)/%.o: $(KERNELS_PATH)/$(1)/%.$(3) $(MK_HEADER_FILES) $(MAKE_DEFS_MK_PATHS)
 ifeq ($(BLIS_ENABLE_VERBOSE_MAKE_OUTPUT),yes)
 	$(CC) $(call get-kernel-cflags-for,$(2)) -c $$< -o $$@
 else
@@ -468,7 +468,9 @@ $(foreach conf, $(CONFIG_LIST), $(eval $(call make-refkern-rule,$(conf))))
 # Instantiate the build rule for optimized kernels for each of the kernel
 # sets in KERNEL_LIST with the CFLAGS designated for the sub-configuration
 # specified by the KCONFIG_MAP.
-$(foreach kset, $(KERNEL_LIST), $(eval $(call make-kernels-rule,$(kset),$(call get-config-for-kset,$(kset)))))
+$(foreach kset, $(KERNEL_LIST), $(eval $(call make-kernels-rule,$(kset),$(call get-config-for-kset,$(kset)),c)))
+$(foreach kset, $(KERNEL_LIST), $(eval $(call make-kernels-rule,$(kset),$(call get-config-for-kset,$(kset)),s)))
+$(foreach kset, $(KERNEL_LIST), $(eval $(call make-kernels-rule,$(kset),$(call get-config-for-kset,$(kset)),S)))
 
 # FGVZ: for later, to compile multiple kernel source suffixes.
 #$(foreach suf,  $(KERNEL_SUFS), \

--- a/config/cortexa15/bli_cntx_init_cortexa15.c
+++ b/config/cortexa15/bli_cntx_init_cortexa15.c
@@ -48,8 +48,8 @@ void bli_cntx_init_cortexa15( cntx_t* cntx )
 	bli_cntx_set_l3_nat_ukrs
 	(
 	  2,
-	  BLIS_GEMM_UKR, BLIS_FLOAT,    bli_sgemm_arm_int_4x4, FALSE,
-	  BLIS_GEMM_UKR, BLIS_DOUBLE,   bli_dgemm_arm_int_4x4, FALSE,
+	  BLIS_GEMM_UKR, BLIS_FLOAT,    bli_sgemm_armv7a_int_4x4, FALSE,
+	  BLIS_GEMM_UKR, BLIS_DOUBLE,   bli_dgemm_armv7a_int_4x4, FALSE,
 	  cntx
 	);
 

--- a/config/cortexa15/bli_family_cortexa15.h
+++ b/config/cortexa15/bli_family_cortexa15.h
@@ -44,14 +44,14 @@
 #if 0
 // -- LEVEL-3 MICRO-KERNEL CONSTANTS -------------------------------------------
 
-#define BLIS_SGEMM_UKERNEL             bli_sgemm_opt_4x4
+#define BLIS_SGEMM_UKERNEL             bli_sgemm_armv7a_int_4x4
 #define BLIS_DEFAULT_MR_S              4
 #define BLIS_DEFAULT_NR_S              4
 #define BLIS_DEFAULT_MC_S              336 
 #define BLIS_DEFAULT_KC_S              528 
 #define BLIS_DEFAULT_NC_S              4096
 
-#define BLIS_DGEMM_UKERNEL             bli_dgemm_opt_4x4
+#define BLIS_DGEMM_UKERNEL             bli_dgemm_armv7a_int_4x4
 #define BLIS_DEFAULT_MR_D              4
 #define BLIS_DEFAULT_NR_D              4
 #define BLIS_DEFAULT_MC_D              176 

--- a/config/cortexa9/bli_cntx_init_cortexa9.c
+++ b/config/cortexa9/bli_cntx_init_cortexa9.c
@@ -48,8 +48,8 @@ void bli_cntx_init_cortexa9( cntx_t* cntx )
 	bli_cntx_set_l3_nat_ukrs
 	(
 	  2,
-	  BLIS_GEMM_UKR, BLIS_FLOAT,    bli_sgemm_arm_int_4x4, FALSE,
-	  BLIS_GEMM_UKR, BLIS_DOUBLE,   bli_dgemm_arm_int_4x4, FALSE,
+	  BLIS_GEMM_UKR, BLIS_FLOAT,    bli_sgemm_armv7a_int_4x4, FALSE,
+	  BLIS_GEMM_UKR, BLIS_DOUBLE,   bli_dgemm_armv7a_int_4x4, FALSE,
 	  cntx
 	);
 

--- a/config/cortexa9/bli_family_cortexa9.h
+++ b/config/cortexa9/bli_family_cortexa9.h
@@ -44,14 +44,14 @@
 #if 0
 // -- LEVEL-3 MICRO-KERNEL CONSTANTS -------------------------------------------
 
-#define BLIS_SGEMM_UKERNEL             bli_sgemm_opt_4x4
+#define BLIS_SGEMM_UKERNEL             bli_sgemm_armv7a_int_4x4
 #define BLIS_DEFAULT_MR_S              4
 #define BLIS_DEFAULT_NR_S              4
 #define BLIS_DEFAULT_MC_S              432
 #define BLIS_DEFAULT_KC_S              352
 #define BLIS_DEFAULT_NC_S              4096
 
-#define BLIS_DGEMM_UKERNEL             bli_dgemm_opt_4x4
+#define BLIS_DGEMM_UKERNEL             bli_dgemm_armv7a_int_4x4
 #define BLIS_DEFAULT_MR_D              4
 #define BLIS_DEFAULT_NR_D              4
 #define BLIS_DEFAULT_MC_D              176

--- a/config/cortexa9/make_defs.mk
+++ b/config/cortexa9/make_defs.mk
@@ -67,7 +67,7 @@ endif
 
 CKOPTFLAGS     := $(COPTFLAGS)
 
-ifneq ($(CC_VENDOR),gcc)
+ifeq ($(CC_VENDOR),gcc)
 CVECFLAGS      := -march=armv7-a
 else
 $(error gcc is required for this configuration.)

